pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,291 @@
+			Name:        "thanos_objstore_azure_storage_operations_total",
+			Help:        "Total number of operations that were executed against an Azure storage account.",
+			ConstLabels: prometheus.Labels{"storage_account": conf.StorageAccountName},
+		}, []string{"operation"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+
+	prefix := dir
+	if prefix != "" && !strings.HasSuffix(prefix, DirDelim) {
+		prefix += DirDelim
+	}
+
+	list, err := b.containerURL.ListBlobsHierarchySegment(ctx, blob.Marker{}, DirDelim, blob.ListBlobsSegmentOptions{
+		Prefix: prefix,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "msg", "cannot list blobs in directory %s", dir)
+	}
+	var listNames []string
+
+	for _, blob := range list.Segment.BlobItems {
+		listNames = append(listNames, blob.Name)
+	}
+
+	for _, blobPrefix := range list.Segment.BlobPrefixes {
+		listNames = append(listNames, blobPrefix.Name)
+	}
+
+	for _, name := range listNames {
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	errorCode := parseError(err.Error())
+	if errorCode == "InvalidUri" || errorCode == "BlobNotFound" {
+		return true
+	}
+
+	return false
+}
+
+func (b *Bucket) getBlobReader(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	level.Debug(b.logger).Log("msg", "Getting blob", "blob", name, "offset", offset, "length", length)
+	if len(name) == 0 {
+		return nil, errors.New("X-Ms-Error-Code: [EmptyContainerName]")
+	}
+	exists, err := b.Exists(ctx, name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "msg", "cannot get blob reader: %s", name)
+	}
+
+	if !exists {
+		return nil, errors.New("X-Ms-Error-Code: [BlobNotFound]")
+	}
+
+	blobURL := getBlobURL(ctx, b.config.StorageAccountName, b.config.StorageAccountKey, b.config.ContainerName, name)
+
+	props, err := blobURL.GetProperties(ctx, blob.BlobAccessConditions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "msg", "cannot get properties for container: %s", name)
+	}
+
+	var size int64
+	if length > 0 {
+		size = length
+	} else {
+		size = props.ContentLength() - offset
+	}
+
+	destBuffer := make([]byte, size)
+
+	if err := blob.DownloadBlobToBuffer(context.Background(), blobURL.BlobURL, offset, length,
+		blob.BlobAccessConditions{}, destBuffer, blob.DownloadFromBlobOptions{
+			BlockSize:   blob.BlobDefaultDownloadBlockSize,
+			Parallelism: uint16(3),
+			Progress:    nil,
+		}); err != nil {
```

Comment:
if you would put `)` at the next line, it would looks nicer.

Reply:
This will break syntax then

Reply:
````
blob.BlobAccessConditions{}, destBuffer, blob.DownloadFromBlobOptions{
			BlockSize:   blob.BlobDefaultDownloadBlockSize,
			Parallelism: uint16(3),
			Progress:    nil,
		},
       ); err != nil {
````
?